### PR TITLE
CPDP-1040: draft pdf submissions with attachments

### PR DESCRIPTION
**Pull request by mikail-khan**

### CODE REVIEW

File: app/uk/gov/hmrc/gform/dms/DmsHtmlSubmission.scala [hunk 1]
```diff
@@ -18,7 +18,7 @@ package uk.gov.hmrc.gform.dms
 
 import play.api.libs.json.{ Format, Json }
 
-case class DmsHtmlSubmission(html: String, metadata: DmsMetadata)
+case class DmsHtmlSubmission(b64html: String, attachments: List[Array[Byte]], metadata: DmsMetadata)
```
Comment (mikail-khan) [inline app/uk/gov/hmrc/gform/dms/DmsHtmlSubmission.scala]: It seems that the html should be sent in b64 format but that was not clear to me until I dug into the code. So making this explicit in the name here.

File: app/uk/gov/hmrc/gform/dms/DmsSubmissionAlgebra.scala [hunk 2]
```diff
@@ -65,7 +67,28 @@ class DmsSubmissionService[F[_]](
...
+    val formTemplateId = FormTemplateId(metaData.dmsFormId)
+
+    for {
+      pdfBytes   <- pdfGenerator.generatePDFBytes(decode(html))
+      envelopeId <- fileUpload.createEnvelope(formTemplateId, LocalDateTime.now(clock).plusDays(formExpiryDays))
+      pdfDoc = documentLoader(pdfBytes)
+      pdfSummary = PdfSummary(pdfDoc.getNumberOfPages.toLong, pdfBytes)
+      _ = pdfDoc.close()
+      submission = DmsSubmissionService
+        .createSubmission(metaData, envelopeId, LocalDateTime.now(clock), Attachments.empty)
```
Comment (mikail-khan) [inline app/uk/gov/hmrc/gform/dms/DmsSubmissionAlgebra.scala]: I went through the code and I still can't see how the attachments are processed. All I can see is that it calculates the number of attachments. What about the content? How is that processed and sent down to DMS with the PDF?

File: app/uk/gov/hmrc/gform/dms/DmsSubmissionWithAttachmentsRequestInterpreter.scala [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+import play.api.libs.json.{ JsError, JsSuccess, Json }
+import play.api.mvc.{ MultipartFormData, Request }
+
+object DmsSubmissionWithAttachmentsRequestInterpreter {
+  def apply(
+    request: Request[MultipartFormData[TemporaryFile]]): Either[String, (String, List[Array[Byte]], DmsMetadata)] = {
+
+    val maybeHtml: Option[Seq[String]] = request.body.dataParts.get("html")
+    val maybeMetadata: Option[DmsMetadata] = Json
+      .toJson(request.body.dataParts.filterKeys(_ == "html").mapValues(_.mkString("")))
```
Comment (mikail-khan) [inline app/uk/gov/hmrc/gform/dms/DmsSubmissionWithAttachmentsRequestInterpreter.scala]: this is a bit messy - I am sending the HTML content and the DMS meta data in the dataparts of the MultipartFormData request.
So I filter out the html to just get the DMS meta keys
if you know a better way to do this, let me know

File: app/uk/gov/hmrc/gform/dms/DmsSubmissionAlgebra.scala (hunk 2 above)
Comment (josef-vlach) [reply app/uk/gov/hmrc/gform/dms/DmsSubmissionAlgebra.scala]: You need to send attachments into `file-upload-frontend`, that logic is not here...

Comment (mikail-khan) [reply app/uk/gov/hmrc/gform/dms/DmsSubmissionAlgebra.scala]: :man_facepalming: 

ok mate, let me check it out and see what I need to do - missing jigsaw in the puzzle.

**Pull request closed without merging**